sqldelight-idea-plugin/src/main/kotlin/com/squareup/sqldelight/intellij/refactoring/SqlDelightSuggestedRefactoringExecution.kt

```diff
@@ -0,0 +1,204 @@
+    val declaration = data.declarationPointer
+    val oldSignature = data.oldSignature
+    val newSignature = data.newSignature
+
+    val file = declaration.containingFile as SqlDelightFile? ?: return null
+
+    val dialect = SqlDelightProjectService.getInstance(file.project).dialectPreset
+    val strategy = SqlGeneratorStrategy.create(dialect)
+
+    val module = ModuleUtil.findModuleForFile(file) ?: return null
+    val fileIndex = SqlDelightFileIndex.getInstance(module)
+    val migrationFile = fileIndex.sourceFolders(file)
+      .flatMap { it.findChildrenOfType<MigrationFile>() }
+      .maxByOrNull { it.version }
+
+    val oldList = oldSignature.parameters
+    val newList = newSignature.parameters
+    val oldColumnDefList = oldList.map { it.columnDef() }
+
+    val tableNameChanged = oldSignature.name != newSignature.name
+    val newColumns = newColumns(oldList, newList)
+    val removedColumns = removedColumns(oldList, newList)
+    val columnNameChanges = columnNameChanges(oldList, newList)
+
+    val migration = mutableListOf<String>()
+    if (tableNameChanged) {
+      migration += strategy.tableNameChanged(oldSignature.name, newSignature.name)
+    }
+    newColumns.forEach { col ->
+      val columnDef = col.columnDef()
+      migration += strategy.columnAdded(oldSignature.name, columnDef)
+    }
+    removedColumns.forEach { col ->
+      migration += strategy.columnRemoved(newSignature.name, col.name, oldColumnDefList)
+    }
+    columnNameChanges.forEach { (oldName, newName) ->
+      migration += strategy.columnNameChanged(newSignature.name, oldName, newName, oldColumnDefList)
+    }
+
+    return SuggestedMigrationData(
+      declarationPointer = declaration,
+      newestMigrationFile = migrationFile,
+      preparedMigration = migration.joinToString("\n").trim()
+    )
+  }
+
+  private fun Parameter.columnDef(): String {
+    val columnConstraints = additionalData as ColumnConstraints?
+    return buildString {
+      append("$name $type")
+      if (columnConstraints != null) {
+        append(" ")
+        append(columnConstraints)
+      }
+    }
+  }
+
+  private fun newColumns(
+    oldList: List<Parameter>,
+    newList: List<Parameter>
+  ): List<Parameter> {
+    return newList.intersectBy(oldList, Parameter::id)
+  }
+
+  private fun removedColumns(oldList: List<Parameter>, newList: List<Parameter>): List<Parameter> {
+    return oldList.intersectBy(newList, Parameter::id)
+  }
+
+  private inline fun <T, R : Any> List<T>.intersectBy(other: List<T>, f: (T) -> R): List<T> {
+    val set = other.mapTo(mutableSetOf(), f)
+    return filter { f(it) !in set }
+  }
+
+  private fun columnNameChanges(
+    oldList: List<Parameter>,
+    newList: List<Parameter>
+  ): List<Pair<String, String>> {
+    val old = oldList.map(Parameter::id)
+    val new = newList.map(Parameter::id)
+    val oldList = oldList.filter { it.id in new }
+    val newList = newList.filter { it.id in old }
```

Comment:
This is quadratic, but it's probably fine because surely people are not getting into thousands of columns... right?

Reply:
Above you use `mapTo(set, function)` to avoid this problem. Any reason not to use it here as well?

Reply:
no reason, just missed it